src/fs/metadata_log_bootstrap.hh

```diff
@@ -0,0 +1,485 @@
+                case CHECKPOINT: // CHECKPOINT cannot appear as part of checkpointed data
+                    return invalid_entry_exception();
+                case NEXT_METADATA_CLUSTER:
+                    return bootstrap_next_metadata_cluster();
+                case CREATE_INODE:
+                    return bootstrap_create_inode();
+                case UPDATE_METADATA:
+                    return bootstrap_update_metadata();
+                case DELETE_INODE:
+                    return bootstrap_delete_inode();
+                case SMALL_WRITE:
+                    return bootstrap_small_write();
+                case MEDIUM_WRITE:
+                    return bootstrap_medium_write();
+                case LARGE_WRITE:
+                    return bootstrap_large_write();
+                case LARGE_WRITE_WITHOUT_MTIME:
+                    return bootstrap_large_write_without_mtime();
+                case TRUNCATE:
+                    return bootstrap_truncate();
+                case MTIME_UPDATE:
+                    return bootstrap_mtime_update();
+                case ADD_DIR_ENTRY:
+                    return bootstrap_add_dir_entry();
+                case CREATE_INODE_AS_DIR_ENTRY:
+                    return bootstrap_create_inode_as_dir_entry();
+                case RENAME_DIR_ENTRY:
+                    return bootstrap_rename_dir_entry();
+                case DELETE_DIR_ENTRY:
+                    return bootstrap_delete_dir_entry();
+                }
+
+                // unknown type => metadata log corruption
+                return invalid_entry_exception();
+            }).then([this] {
+                if (_curr_checkpoint.bytes_left() > 0) {
+                    return invalid_entry_exception(); // Corrupted checkpointed data
+                }
+                return now();
+            });
+        });
+    }
+
+    future<> bootstrap_next_metadata_cluster() {
+        ondisk_next_metadata_cluster entry;
+        if (not _curr_checkpoint.read_entry(entry)) {
+            return invalid_entry_exception();
+        }
+
+        _next_cluster = (cluster_id_t)entry.cluster_id;
+        return now();
+    }
+
+    bool inode_exists(inode_t inode) {
+        return _metadata_log._inodes.count(inode) != 0;
+    }
+
+    future<> bootstrap_create_inode() {
+        ondisk_create_inode entry;
+        if (not _curr_checkpoint.read_entry(entry) or inode_exists(entry.inode)) {
+            return invalid_entry_exception();
+        }
+
+        _metadata_log.memory_only_create_inode(entry.inode, entry.is_directory, ondisk_metadata_to_metadata(entry.metadata));
+        return now();
+    }
+
+    future<> bootstrap_update_metadata() {
+        ondisk_update_metadata entry;
+        if (not _curr_checkpoint.read_entry(entry) or not inode_exists(entry.inode)) {
+            return invalid_entry_exception();
+        }
+
+        _metadata_log._inodes[entry.inode].metadata = ondisk_metadata_to_metadata(entry.metadata);
```

Comment:
Maybe in order to be consistent use something like _metadata_log.update_metadata(entry.inode, entry.metadata).